### DISCUSSION TITLE
limit concurrent connections for one Channel

### DISCUSSION THREAD

**Discussion by rawezhcode** [centrifugal/centrifugo#815] [Q&A]
What is the limit for one channel?

I have a Q&A app that starts live at a certain time at night .
  For example, the game starts at 10 pm with more than 10k users.
  I'm currently having trouble in real-time,
my app that is made with (flutter),
 I only use (broadcast) to send questions to users.
What vps is good for a centrifuge that can accept 10k concurrent connections in one Channel ?

**Comment by FZambia** [comment]
You must measure yourself, it's impossible to say without knowing all the details.

**Reply by rawezhcode** [reply]
I have an app with about 9k users and I'm currently using laravel reverb to send questions to participants, but there are some problems, so I want to try centrifugo.
I have two servers, one with my app and the other with realtime.

VPS 1: database and app (in Atlantic) / CPU: 6 vCPU, RAM: 16 GB
VPS 2: Rrealtime (in Atlantic) / CPU: 8 vCPU, RAM: 32 GB

**Reply by FZambia** [reply]
I believe Centrifugo should be capable to serve no less than 100k on such hardware in idle mode (limiting factor will be CPU) - but there are many other things involved – message rate, message size, features enabled which can affect resource usage. That's why it's hard to say exactly. I suggest to try. Will be very curious to listen to your results.

**Reply by rawezhcode** [reply]
Very well so testing is the best thing I can do .
 So I'll try to see how much it will help me.

Note: 
I send 3 messages to all users every minute,
Each message size is about 120kb.

**Reply by FZambia** [reply]
I believe it should be simple to achieve with Centrifugo - check out this blog post https://centrifugal.dev/blog/2020/02/10/million-connections-with-centrifugo – probably it will provide you some insights about hardware.

Also, see https://centrifugal.dev/docs/tutorial/scale where one message was broadcasted to 10k subscribers with different channels, in your case it should be even faster if there is a single channel for all 9k users.

**Reply by rawezhcode** [reply]
Thanks.

Is there any limit to the presence Channel,
Is Presence Channel capable of 9k users? Or should I use a private channel?

**Reply by FZambia** [reply]
Check out this - https://centrifugal.dev/docs/faq#how-scalable-is-the-online-presence-and-joinleave-features
And this - https://centrifugal.dev/docs/getting-started/design#online-presence-considerations

In general, depends how you are going to use presence. If it's for backend needs for rare requests – then it may be OK. But it still may block Redis for some time (need to measure actually, but must be rather fast in Memory engine case). If you are going to ask presence in channels with 9k online subscribers from the client side - it's a bad idea. Also, join/leave messages are very expensive in channels of this size - because every join or leave must be broadcasted to every member - it's n^2 complexity. Also take a look at [presence stats](https://centrifugal.dev/docs/server/server_api#presence_stats) feature - just 2 counters and it's much more efficient - maybe it's enough.

**Reply by rawezhcode** [reply]
./centrifugo gensubtoken --user rawezhcode --channel presence-public:chats -t -1

Using this code to create tokens for users,
Does it cause me any problems in terms of the number of users ?

**Reply by FZambia** [reply]
By default it should not, there is an option https://centrifugal.dev/docs/server/configuration#client_user_connection_limit - but it's not used by default.
